### PR TITLE
Torch Elastic DDP DeadLock bug fix

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduces some changes to better support multi node dead lock mechanism prevention when processes are created by the cluster.

This works.
```
MASTER_ADDR="localhost" MASTER_PORT=1234 WORLD_SIZE=2 LOCAL_RANK=0 python tests/plugins/environments/torch_elastic_deadlock.py
MASTER_ADDR="localhost" MASTER_PORT=1234 WORLD_SIZE=2 LOCAL_RANK=1 python tests/plugins/environments/torch_elastic_deadlock.p
```

TorchElastic detects rank 1 as Terminated as expected. Therefore, rank 0 has succeeded and rank 1 failed. Therefore the run is declared as failed. Depending on use-case, this can be a wrong behaviour.

```
python -m torch.distributed.run --nproc_per_node=2 --max_restarts 0 tests/plugins/environments/torch_elastic_deadlock.py
```

TorchElastic should expose a kill function, so it registers that we manual kill a process.
```
from torch.distributed.elastic.utils import elatic_kill
```

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -450,7 +448,6 @@ def reconciliate_processes(self, trace: str):
 
         for pid in self._pids:
             if pid != os.getpid():
-                os.kill(pid, signal.SIGTERM)
-                # os.kill(pid, signal.SIGKILL)
-            shutil.rmtree(sync_dir)
-            raise DeadlockDetectedException(f"DeadLock detected from rank: {self.global_rank} \n {trace}")
+                os.kill(pid, signal.SIGKILL)
+        shutil.rmtree(sync_dir)
+        raise DeadlockDetectedException(f"DeadLock detected from rank: {self.global_rank} \n {trace}")
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:453]: do we have the prior exception context to `raise Exception from e` ?

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:453]: Yes, it is passed as the trace.

**Comment by tchaton** [comment]
Hey @ananthsub.

I would add a Multi Node test running on Grid using Torch Elastic soon, so our support for Torch Elastic is high.

Best,
T.C

**Review by awaelchli — APPROVED** [review]

File: tests/special_tests.sh [hunk 2]
```diff
@@ -78,6 +78,12 @@ if [ $? -eq 0 ]; then
     report+="Ran\ttests/utilities/test_warnings.py\n"
 fi
 
+# test deadlock is properly handled with TorchElastic.
+LOGS=$(PL_RUNNING_SPECIAL_TESTS=1 python -m torch.distributed.launch --nproc_per_node=2 --max_restarts 0 -m coverage run --source pytorch_lightning -a tests/plugins/environments/torch_elastic_deadlock.py | grep "SUCCEEDED")
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/special_tests.sh:82]: The CI still runs on 1.8 and not 1.9. I believe run was introduced later on.

File: tests/plugins/environments/torch_elastic_deadlock.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    trainer = Trainer(
+        default_root_dir=".", max_epochs=1, limit_train_batches=5, num_sanity_val_steps=0, gpus=2, accelerator="ddp"
+    )
+    assert isinstance(trainer.training_type_plugin, DDPPlugin)
+
+    try:
+        # simulate random failure in training_step on rank 0
+        trainer.fit(model)
+    except Exception:
+        pass
```
Comment (carmocca) [inline tests/plugins/environments/torch_elastic_deadlock.py:32]: Can you assert the expected exception type?

Is this for `CustomException`? Where is the `DeadlockDetectedException` handled?

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 4]
```diff
@@ -410,8 +407,18 @@ def register_plugins(cls, plugin_registry: Dict) -> None:
...
+        # there should be a unique sync_dir per nodes.
+        if self.local_rank == 0:
+            # create a temporary directory used to synchronize processes on deadlock.
+            self._sync_dir = tempfile.mkdtemp()
+
+        sync_dirs = []
+        node_zero = 0
+        for _ in range(self.num_nodes):
+            sync_dirs.append(self.broadcast(self._sync_dir, node_zero))
+            node_zero += self.world_size // self.num_nodes
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp.py:419]: noob q: what is this calculating?

also, wouldn't it equal

```suggestion
    node_zero = (self.world_size // self.num_nodes) * self.num_nodes
```

File: tests/special_tests.sh [hunk 5]
```diff
@@ -78,6 +78,12 @@ if [ $? -eq 0 ]; then
...
+if  [ -z "$LOGS" ]; then
+    exit 1
+fi
```
Comment (carmocca) [inline tests/special_tests.sh:85]:
```suggestion
fi
report+="Ran\ttests/plugins/environments/torch_elastic_deadlock.py\n"
```

**Review by carmocca — APPROVED** [review]

File: tests/plugins/environments/torch_elastic_deadlock.py (hunk 3 above)
Comment (tchaton) [reply tests/plugins/environments/torch_elastic_deadlock.py:32]: Done !

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:419]: It is getting the indices of each first rank of each node and performing a broadcast of its local sync_dir

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py:419]: Example: 4 nodes with 8 gpus: [0, 8, 16, 24]

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py]: @carmocca

**Review by SeanNaren — APPROVED** [review]
More important to get this in even if there is no test atm I think

File: tests/plugins/environments/torch_elastic_deadlock.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import os
```
Comment (ananthsub) [inline tests/plugins/environments/torch_elastic_deadlock.py:1]: nit on test location: this isn't really torchelastic specific
it's more about how the DDP plugin handles deadlocks when processes are created externally to the plugin
maybe we could add it here: under `test_ddp_plugin_deadlock_detection` ? 
https://github.com/PyTorchLightning/pytorch-lightning/tree/master/tests/plugins

Comment (tchaton) [reply tests/plugins/environments/torch_elastic_deadlock.py:1]: I will make another PR with this update. GPU CI has been flicky today + I want to give a try to multi node and torch elastic.
